lib/nrf_modem_lib/nrf_modem_lib_trace.c

```diff
@@ -88,38 +191,35 @@ int nrf_modem_lib_trace_start(enum nrf_modem_lib_trace_mode trace_mode)
 
 int nrf_modem_lib_trace_process(const uint8_t *data, uint32_t len)
 {
-	if (!is_transport_initialized) {
-		return -ENXIO;
-	}
+	int ret = 0;
 
-#ifdef CONFIG_NRF_MODEM_LIB_TRACE_MEDIUM_UART
-	/* Split RAM buffer into smaller chunks to be transferred using DMA. */
-	uint32_t remaining_bytes = len;
-	const uint32_t MAX_BUF_LEN = (1 << UARTE1_EASYDMA_MAXCNT_SIZE) - 1;
+	if (is_transport_initialized) {
+		struct trace_data_t trace_data = { .data = data, .len = len };
+		char *mem_ptr = k_malloc(sizeof(trace_data));
 
-	while (remaining_bytes) {
-		size_t transfer_len = MIN(remaining_bytes, MAX_BUF_LEN);
-		uint32_t idx = len - remaining_bytes;
+		if (mem_ptr != NULL) {
+			memcpy(mem_ptr, &trace_data, sizeof(trace_data));
 
-		nrfx_uarte_tx(&uarte_inst, &data[idx], transfer_len);
-		remaining_bytes -= transfer_len;
-	}
-#endif
+			k_fifo_put(&trace_fifo, mem_ptr);
+		} else {
+			LOG_ERR("k_malloc failed.");
```

Comment:
When testing with the application_update example, heap memory allocation for the trace fifo queue fails very often when `CONFIG_HEAP_MEM_POOL_SIZE` is 2048 (the example default). After increasing `CONFIG_HEAP_MEM_POOL_SIZE` to somewhere around 3328, heap allocations no longer fails. Example modem_delta_update behaves similar.

I guess this is to be expected and that the user will have to manually increase the size of the heap memory pool to a suitable value when enabling modem trace. Was thinking if the default `CONFIG_HEAP_MEM_POL_SIZE` of the examples could be increased to avoid this heap memory allocation failures, but I realize that might be undesirable in the case where you do not have modem trace enabled.

Should we look at this as an issue or intended behavior?

Reply:
When looking at the different samples many of them already have 4k and higher `CONFIG_HEAP_MEM_POOL_SIZE`. At least those involving FOTA updates. So I'd argue it wouldn't hurt to increase `CONFIG_HEAP_MEM_POOL_SIZE` in the application_update sample to 4k even without traces enabled.

I can update the docs and changelog, and I can mention the use of heap there.

Reply:
Docs and changelog updated

Reply:
It would probably be better to split the library heap in two, so that tracing can have its dedicated heap space.
As it is now, there is an undesirable relation between how fast are traces processed/freed and the other functionality that depends on heap space, such as networking sockets. It would be best if they were isolated, so that the application does not need to change the library heap size based on whether tracing is on, and so that runtime tracing performance does not affect the rest of the library.

Reply:
Note that the heap used in the piece of code is the application heap and not the one defined by `CONFIG_NRF_MODEM_LIB_HEAP_SIZE`. So I believe your comment is about `CONFIG_NRF_MODEM_LIB_HEAP_SIZE` and how we have increased when traces are enabled. If thats the case, then yes. It would be nice to have two separate heaps,  one for traces and one more other operations. That would mean that the `nrf_modem_os_alloc()` should be made aware of which heap to use when the libmodem requests for heap space. Is that acceptable? Or is there a better solution?

Reply:
Me and @gregersrygg  talked about it. It is a good idea to have a separate "trace heap". That means both this thread and the libmodem will allocate from the same 'trace heap'. Perhaps this can be done in a separate PR(s) in the near future by the libmodem team.